File: lib/test/t_ComposedCopula_std.cxx
//                                               -*- C++ -*-
/**
 *  @file  t_ComposedCopula_std.cxx
 *  @brief The test file of class ComposedCopula for standard methods
 *
 *  Copyright (C) 2005-2013 EDF-EADS-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 *  @author schueller
 *  @date   2012-07-16 15:59:45 +0200 (Mon, 16 Jul 2012)
 */
#include "OT.hxx"
#include "OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

int main(int argc, char *argv[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);
  setRandomGenerator();

  try
    {
      // Instanciate one distribution object
      CorrelationMatrix R(3);
      R(0, 1) = 0.5;
      R(0, 2) = 0.25;
      ComposedCopula::CopulaCollection collection(3);
      collection[0] = FrankCopula(3.0);
      collection[1] = NormalCopula(R);
      collection[2] = ClaytonCopula(2.0);
      ComposedCopula copula(collection);
      fullprint << "Copula " << copula << std::endl;
      std::cout << "Copula " << copula << std::endl;

      // Is this copula elliptical ?
      fullprint << "Elliptical distribution= " << (copula.isElliptical() ? "true" : "false") << std::endl;

      // Is this copula continuous ?
      fullprint << "Continuous = " << (copula.isContinuous() ? "true" : "false") << std::endl;

      // Is this copula elliptical ?
      fullprint << "Elliptical = " << (copula.hasEllipticalCopula() ? "true" : "false") << std::endl;

      // Is this copula independent ?
      fullprint << "Independent = " << (copula.hasIndependentCopula() ? "true" : "false") << std::endl;

      // Test for realization of copula
      NumericalPoint oneRealization = copula.getRealization();
      fullprint << "oneRealization=" << oneRealization << std::endl;

      // Test for sampling
      UnsignedLong size = 10000;
      NumericalSample oneSample = copula.getSample( size );
      fullprint << "oneSample first=" << oneSample[0] << " last=" << oneSample[size - 1] << std::endl;
      fullprint << "mean=" << oneSample.computeMean() << std::endl;
      UnsignedLong precision(PlatformInfo::GetNumericalPrecision());
      PlatformInfo::SetNumericalPrecision(5);
      fullprint << "covariance=" << oneSample.computeCovariance() << std::endl;
      PlatformInfo::SetNumericalPrecision(precision);

      // Define a point
      NumericalPoint point( copula.getDimension(), 0.6 );
      fullprint << "Point= " << point << std::endl;

      // Show PDF and CDF of point
      NumericalScalar eps(1e-5);
      NumericalPoint DDF = copula.computeDDF( point );
      fullprint << "ddf     =" << DDF << std::endl;
      NumericalPoint ddfFD(copula.getDimension());
      fullprint << "ddf (FD)=" << copula.ContinuousDistribution::computeDDF(point) << std::endl;
      NumericalScalar PDF = copula.computePDF( point );
      fullprint << "pdf     =" << PDF << std::endl;
      NumericalScalar CDF = copula.computeCDF( point );
      fullprint << "cdf=" << CDF << std::endl;
      //    NumericalPoint PDFgr = copula.computePDFGradient( point );
      //    fullprint << "pdf gradient     =" << PDFgr << std::endl;
      //    NumericalPoint CDFgr = copula.computeCDFGradient( point );
      //    fullprint << "cdf gradient     =" << CDFgr << std::endl;
      NumericalPoint quantile = copula.computeQuantile( 0.95 );
      fullprint << "quantile=" << quantile << std::endl;
      fullprint << "cdf(quantile)=" << copula.computeCDF(quantile) << std::endl;
      NumericalPoint mean = copula.getMean();
      fullprint << "mean=" << mean << std::endl;
      CovarianceMatrix covariance = copula.getCovariance();
      precision = PlatformInfo::GetNumericalPrecision();
      PlatformInfo::SetNumericalPrecision(5);
      fullprint << "covariance=" << covariance << std::endl;
      PlatformInfo::SetNumericalPrecision(precision);
      ComposedCopula::NumericalPointWithDescriptionCollection parameters = copula.getParametersCollection();
      fullprint << "parameters=" << parameters << std::endl;

      // Specific to this copula

      // Extract a 5-D marginal
      UnsignedLong dim(5);
      point = NumericalPoint(dim, 0.25);
      Indices indices(dim, 0);
      indices[0] = 1;
      indices[1] = 2;
      indices[2] = 3;
      indices[3] = 5;
      indices[4] = 6;
      fullprint << "indices=" << indices << std::endl;
      Copula margins(copula.getMarginal(indices));
      fullprint << "margins=" << margins << std::endl;
      fullprint << "margins PDF=" << margins.computePDF(point) << std::endl;
      fullprint << "margins CDF=" << margins.computeCDF(point) << std::endl;
      quantile = margins.computeQuantile(0.95);
      fullprint << "margins quantile=" << quantile << std::endl;
      fullprint << "margins CDF(quantile)=" << margins.computeCDF(quantile) << std::endl;
      fullprint << "margins realization=" << margins.getRealization() << std::endl;
    }
  catch (TestFailed & ex)
    {
      std::cerr << ex << std::endl;
      return ExitCode::Error;
    }


  return ExitCode::Success;
}

File: python/test/t_ComposedCopula_std.py
#! /usr/bin/env python

from openturns import *

TESTPREAMBLE()
RandomGenerator().SetSeed(0)

try :
    # Instanciate one distribution object
    R = CorrelationMatrix(3)
    R[0, 1] = 0.5
    R[0, 2] = 0.25
    copula = ComposedCopula([FrankCopula(3.0), NormalCopula(R), ClaytonCopula(2.0)])

    print "Copula ", repr(copula)
    print "Copula ", copula

    # Is this copula elliptical ?
    print "Elliptical distribution= ", copula.isElliptical()

    # Is this copula continuous ?
    print "Continuous = ", copula.isContinuous()

    # Is this copula elliptical ?
    print "Elliptical = ", copula.hasEllipticalCopula()

    # Is this copula independent ?
    print "Independent = ", copula.hasIndependentCopula()

    # Test for realization of copula
    oneRealization = copula.getRealization()
    print "oneRealization=", repr(oneRealization)

    # Test for sampling
    size = 10000
    oneSample = copula.getSample( size )
    print "oneSample first=", repr(oneSample[0]), " last=", repr(oneSample[size - 1])
    print "mean=", repr(oneSample.computeMean())
    print "covariance=", repr(oneSample.computeCovariance())

    # Define a point
    point = NumericalPoint( copula.getDimension(), 0.6 )
    print "Point= ", repr(point)

    # Show PDF and CDF of point
    # NumericalScalar eps(1e-5)
    DDF = copula.computeDDF( point )
    print "ddf     =", repr(DDF)
    PDF = copula.computePDF( point )
    print "pdf     =%.6f" % PDF
    CDF = copula.computeCDF( point )
    print "cdf=%.6f" % CDF
    #    NumericalPoint PDFgr = copula.computePDFGradient( point )
    #    print "pdf gradient     =", PDFgr
    #    NumericalPoint CDFgr = copula.computeCDFGradient( point )
    #    print "cdf gradient     =", CDFgr
    quantile = copula.computeQuantile( 0.95 )
    print "quantile=", repr(quantile)
    print "cdf(quantile)=%.6f" % copula.computeCDF(quantile)
    mean = copula.getMean()
    print "mean=", repr(mean)
    covariance = copula.getCovariance()
    print "covariance=", repr(covariance)
    parameters = copula.getParametersCollection()
    print "parameters=", repr(parameters)

    # Specific to this copula

    # Extract a 5-D marginal
    dim = 5
    point = NumericalPoint(dim, 0.25)
    indices = Indices(dim, 0)
    indices[0] = 1
    indices[1] = 2
    indices[2] = 3
    indices[3] = 5
    indices[4] = 6
    print "indices=", repr(indices)
    margins = copula.getMarginal(indices)
    print "margins=", repr(margins)
    print "margins PDF=%.6f" % margins.computePDF(point)
    print "margins CDF=%.6f" % margins.computeCDF(point)
    quantile = margins.computeQuantile(0.95)
    print "margins quantile=", repr(quantile)
    print "margins CDF(quantile)=%.6f" % margins.computeCDF(quantile)
    print "margins realization=", repr(margins.getRealization())

except :
    import sys
    print "t_ComposedCopula.py", sys.exc_type, sys.exc_value
